subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java

```diff
@@ -55,15 +56,21 @@
 public abstract class AbstractFileCollection implements FileCollectionInternal {
     protected final TaskDependencyFactory taskDependencyFactory;
     protected final Factory<PatternSet> patternSetFactory;
+    protected final FileCollectionListener fileCollectionListener;
 
-    protected AbstractFileCollection(TaskDependencyFactory taskDependencyFactory, Factory<PatternSet> patternSetFactory) {
+    protected AbstractFileCollection(
+        TaskDependencyFactory taskDependencyFactory,
+        Factory<PatternSet> patternSetFactory,
+        FileCollectionListener fileCollectionListener
+    ) {
         this.taskDependencyFactory = taskDependencyFactory;
         this.patternSetFactory = patternSetFactory;
+        this.fileCollectionListener = fileCollectionListener;
     }
 
     @SuppressWarnings("deprecation")
-    protected AbstractFileCollection(TaskDependencyFactory taskDependencyFactory)  {
-        this(taskDependencyFactory, PatternSets.getNonCachingPatternSetFactory());
+    protected AbstractFileCollection(TaskDependencyFactory taskDependencyFactory) {
+        this(taskDependencyFactory, PatternSets.getNonCachingPatternSetFactory(), Instrumented::fileCollectionObserved);
```

Comment:
As opposite - we could invoke static method of `Instrumented`. The downside is adding dependency `:core` to `:file-collection` what is makes cycle. As a solution we could move `Instrumented` to `:core-api` but this also drags a bunch of files for moving